core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -83,83 +84,200 @@ public static SortedSet<String> listSchemas(Session session, Metadata metadata,
 
     public static SortedSet<String> listSchemas(Session session, Metadata metadata, AccessControl accessControl, String catalogName, Optional<String> schemaName)
     {
-        Set<String> schemaNames = ImmutableSet.copyOf(metadata.listSchemaNames(session, catalogName));
-        if (schemaName.isPresent()) {
-            // we don't use metadata.schemaExists(), because this would change semantics of the method (all vs visible schemas)
-            if (!schemaNames.contains(schemaName.get())) {
-                return ImmutableSortedSet.of();
+        try {
+            Set<String> schemaNames = ImmutableSet.copyOf(metadata.listSchemaNames(session, catalogName));
+            if (schemaName.isPresent()) {
+                // we don't use metadata.schemaExists(), because this would change semantics of the method (all vs visible schemas)
+                if (!schemaNames.contains(schemaName.get())) {
+                    return ImmutableSortedSet.of();
+                }
+                schemaNames = ImmutableSet.of(schemaName.get());
             }
-            schemaNames = ImmutableSet.of(schemaName.get());
+            return ImmutableSortedSet.copyOf(accessControl.filterSchemas(session.toSecurityContext(), catalogName, schemaNames));
+        }
+        catch (RuntimeException exception) {
+            if (exception instanceof TrinoException trinoException) {
```

Comment:
I don't think we need this condition. Let's simply add this exception as cause.

Reply:
Do you simply mean to write it like this (rather than `instanceof` check)?
```java
try {
    ...
catch (TrinoException e) {
    ...
}
catch (RuntimeException e) {
    ...
}
```
Or did you mean to attach it as a suppression and leave the original alone? In which case we can simplify the code like this:
```java
try {
    ...
}
catch (RuntimeException e) {
    e.addSuppressed(new RuntimeException("Catalog: " + catalogName));
    throw e;
}
```
I like the simplicity, but this might not have the desired effect of readily surfacing the error, depending on the tool being used.

Reply:
Turns out some connectors usually throw `RuntimeException`, so I changed to always add the original message.